comp/api/api/utils/common.go

```diff
@@ -0,0 +1,107 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023-present Datadog, Inc.
+
+// Package utils has common utility methods that components can use for structuring http responses of their endpoints
+package utils
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/logs/diagnostic"
+	"github.com/DataDog/datadog-agent/pkg/util/grpc"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// SetJSONError writes a server error as JSON with the correct http error code
+func SetJSONError(w http.ResponseWriter, err error, errorCode int) {
+	w.Header().Set("Content-Type", "application/json")
+	body, _ := json.Marshal(map[string]string{"error": err.Error()})
+	http.Error(w, string(body), errorCode)
+}
+
+// GetConnection returns the connection for the request
+func GetConnection(r *http.Request) net.Conn {
+	return r.Context().Value(grpc.ConnContextKey).(net.Conn)
+}
+
+// MessageReceiver is an exported interface for a valid receiver of streamed output
+type MessageReceiver interface {
+	SetEnabled(e bool) bool
+	Filter(filters *diagnostic.Filters, done <-chan struct{}) <-chan string
+}
+
+// GetStreamFunc returns a handlerfunc that handles request to stream output to the desired receiver
+func GetStreamFunc(messageReceiverFunc func() MessageReceiver, streamType, agentType string) func(w http.ResponseWriter, r *http.Request) {
```

Comment:
This functions and the two new imports `github.com/DataDog/datadog-agent/pkg/util/log` and `"github.com/DataDog/datadog-agent/pkg/logs/diagnostic"` is adding +40 packages to some agent binaries.

The first thing I can think of is to not use the `pkg/util/log` but to use the component core log component which would remove all the imports belonging to `pkg/util/log` and it has the same functionality 

The second thing I would do since we can not remove `github.com/DataDog/datadog-agent/pkg/logs/diagnostic` entirely is create a `api/utils/stream` package and move `MessageReceiver` and the `GetStreamFunc` to it, that way only the places that need stream HTTP response would require that package reducing the total import packages.

Reply:
I think the second approach is great. Ideally, we should avoid having `common` package when few functions bring dependencies.

Reply:
This has been fixed in https://github.com/DataDog/datadog-agent/pull/25367/commits/43e85bb9f06fd8da90f9bf67113edb70f2e9b2a3

and you can see the updated [import list](https://github.com/DataDog/datadog-agent/pull/25367#issuecomment-2096039811) only has a +1 import for the subpackage

Reply:
That is amazing @jeremy-hanna 

One note. We still use the `github.com/DataDog/datadog-agent/pkg/util/log` inside the new `comp/api/api/utils/stream/stream.go`. I thihnk it would be better to use the log component that the pkg version. What do you think?